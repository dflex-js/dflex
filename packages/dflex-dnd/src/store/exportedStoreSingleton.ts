import DFlexDnDExportedStore from "./DFlexDnDExportedStore";

function createStoreInstance() {
  const store = DFlexDnDExportedStore.getInstance();

  return store;
}

export default createStoreInstance();
