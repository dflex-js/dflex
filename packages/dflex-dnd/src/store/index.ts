// eslint-disable-next-line import/prefer-default-export
export { default as exportedStoreSingleton } from "./exportedStoreSingleton";
export { default as store } from "./storeSingleton";
