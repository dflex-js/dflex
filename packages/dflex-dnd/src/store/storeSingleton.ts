import { canUseDOM } from "@dflex/utils";
import { DFlexDnDStore } from "../LayoutManager";

declare global {
  // eslint-disable-next-line
  var $DFlex: DFlexDnDStore;
}

function createStoreInstance() {
  const store = DFlexDnDStore.getInstance();

  if (__DEV__) {
    if (canUseDOM()) {
      if (!globalThis.$DFlex) {
        globalThis.$DFlex = store;
      }
    }
  }

  return store;
}

export default createStoreInstance();
