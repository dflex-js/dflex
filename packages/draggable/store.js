import CoreInstance from "@dflex/core-instance/src";
import AbstractStore from "@dflex/store/src/AbstractStore";

class DraggableStoreImp extends AbstractStore {
  register(elmInstance) {
    super.register(elmInstance, CoreInstance);
  }
}

let store;

// eslint-disable-next-line func-names
export default (function () {
  if (!store) {
    store = new DraggableStoreImp();
  }

  return store;
})();
