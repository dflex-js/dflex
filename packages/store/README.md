<h1 align="center">
  <img
  src="https://raw.githubusercontent.com/jalal246/dflex/master/DFlex-readme.png"
  alt="Dflex logo" />
</h1>

<p align="center">
  <a href="https://github.com/dflex-js/dflex">
    <img
    src="https://img.shields.io/github/workflow/status/dflex-js/dflex/Unit Test"
    alt="Dflex build status" />
  </a>
  <a href="https://github.com/dflex-js/dflex/pulls">
    <img
    src="https://img.shields.io/github/issues-pr/dflex-js/dflex"
    alt="number of opened pull requests"/>
  </a>
  <a href="https://github.com/dflex-js/dflex/issues">
  <img
    src="https://img.shields.io/github/issues/dflex-js/dflex"
    alt="number of opened issues"/>
  </a>
  <a href="https://github.com/dflex-js/dflex/pulls">
   <img
   src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg"
   alt="Dflex welcomes pull request" />
  </a>
</p>

# @dflex/store

> Traverse through the DOM tree using element-id.

DFlex store works on connecting all registered elements and organizes the
relations between them using the DOM relations generator algorithm. Allowing
each element to be reached recursively using its id.

## Installation

```bash
npm install @dflex/store
```

## API

Visit DFlex site for more <https://www.dflex.dev/> and to see live
examples with the full code.
