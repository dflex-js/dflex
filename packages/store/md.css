img[src*="#center"] {
  display: block;
  margin: auto;
}
