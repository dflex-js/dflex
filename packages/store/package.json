{
  "name": "@dflex/store",
  "version": "1.3.0",
  "main": "dist/dflexStore.js",
  "author": "Jalal Maskoun <jimmy002020@gmail.com>",
  "scripts": {
    "types": "tsc --emitDeclarationOnly",
    "prebuild": "yarn types",
    "build": "esbuild src/index.ts --format=esm --bundle --outfile=dist/dflexStore.js"
  },
  "homepage": "https://github.com/jalal246/dflex/tree/master/packages/store",
  "repository": "https://github.com/jalal246/dflex",
  "license": "GPL-3.0",
  "files": [
    "dist",
    "LICENSE"
  ],
  "bundledDependencies": [
    "@dflex/dom-gen"
  ],
  "keywords": [
    "@dflex",
    "@dflex/dom-gen",
    "@dflex/store",
    "@dflex/core-instance",
    "@dflex/dnd",
    "@folo",
    "builderz",
    "DOM",
    "front-end"
  ],
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  }
}
