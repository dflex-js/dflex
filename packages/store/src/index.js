import store from "./init";

export default store;
