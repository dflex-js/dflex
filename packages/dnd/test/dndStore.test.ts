import store from "../src/DnDStore";
// import DraggableInteractive from "../src/Draggable";
// import DistanceCalculator from "../src/Droppable/DistanceCalculator";
// import extractOpts from "../src/utils/extractOpts";

import { elm1, elm2, elm3, elm4 } from "./__mocks__/DOMElements";

describe("DnD Store", () => {
  beforeAll(() => {
    store.register(elm1);
  });

  describe("Shallow test the store", () => {
    it("Element is initiated", () => {
      store.register(elm1);
      expect(store.registry).toMatchSnapshot();
    });

    it("Returns element branch and parent", () => {
      const { branches, parent } = store.getElmTreeById(elm1.id);

      expect(branches).toMatchInlineSnapshot(`
      Object {
        "parents": undefined,
        "siblings": Array [
          "id-1",
        ],
      }
    `);

      expect(parent).toMatchInlineSnapshot(`null`);
    });
  });

  describe("getInsertionELmMeta", () => {
    const preservePosition = {
      x: 1000,
      y: 1000,
    };

    let SK: string;

    describe("Normal branch without injection", () => {
      it("Checks orphan branch", () => {
        ({
          element: {
            keys: { SK },
          },
        } = store.getElmTreeById(elm1.id));

        const { elm, ...rest } = store.getInsertionELmMeta(0, SK);

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": true,
                    "isRestoredLastPosition": false,
                    "position": PointNum {
                      "x": 450,
                      "y": 114,
                    },
                    "prevElm": null,
                  }
              `);
      });

      it("Register three elements in the store", () => {
        store.register(elm2);
        store.register(elm3);
        store.register(elm4);

        expect(store.DOMGen.branches[SK]).toMatchInlineSnapshot(`
          Array [
            "id-1",
            "id-2",
            "id-3",
            "id-4",
          ]
        `);
      });

      it("Checks the last element", () => {
        const { elm, prevElm, position, ...rest } = store.getInsertionELmMeta(
          3,
          SK
        );

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": false,
                  }
              `);

        const pos = elm4.ref.getBoundingClientRect();

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm.id).toBe(elm4.id);
        expect(prevElm.id).toBe(elm3.id);
      });

      it("Preserve the last element", () => {
        store.containers[SK].preservePosition(preservePosition);
      });

      it("Restores the preserved position when calling for last element", () => {
        const { elm, prevElm, position, ...rest } = store.getInsertionELmMeta(
          3,
          SK
        );

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        expect(position.getInstance()).toStrictEqual(preservePosition);

        expect(elm.id).toBe(elm4.id);
        expect(prevElm.id).toBe(elm3.id);
      });

      it("Restores the position even preserved is defined because it's not the last element", () => {
        const { elm, prevElm, position, ...rest } = store.getInsertionELmMeta(
          2,
          SK
        );

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = elm3.ref.getBoundingClientRect();

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm.id).toBe(elm3.id);
        expect(prevElm.id).toBe(elm2.id);
      });

      it("Restores the position for the first element with correct flags", () => {
        const { elm, prevElm, position, ...rest } = store.getInsertionELmMeta(
          0,
          SK
        );

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = elm1.ref.getBoundingClientRect();

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm.id).toBe(elm1.id);
        expect(prevElm).toBeNull();
      });
    });

    describe("Injected branch with empty string", () => {
      it("Inject empty string", () => {
        store.DOMGen.branches[SK].push("");

        expect(store.DOMGen.branches[SK]).toMatchInlineSnapshot(`
          Array [
            "id-1",
            "id-2",
            "id-3",
            "id-4",
            "",
          ]
        `);
      });

      it("Restores the preserved position when calling for last element", () => {
        const { elm, prevElm, position, ...rest } = store.getInsertionELmMeta(
          4,
          SK
        );

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        expect(position.getInstance()).toStrictEqual(preservePosition);

        expect(elm.id).toBe(elm4.id);
        expect(prevElm.id).toBe(elm3.id);
      });

      it("Restores the position even preserved is defined because it's not the last element", () => {
        const { elm, prevElm, position, ...rest } = store.getInsertionELmMeta(
          3,
          SK
        );

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = elm4.ref.getBoundingClientRect();

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm.id).toBe(elm4.id);
        expect(prevElm.id).toBe(elm3.id);
      });

      it("Restores the position for the first element with correct flags", () => {
        const { elm, prevElm, position, ...rest } = store.getInsertionELmMeta(
          0,
          SK
        );

        expect(rest).toMatchInlineSnapshot(`
                  Object {
                    "isEmpty": false,
                    "isOrphan": false,
                    "isRestoredLastPosition": true,
                  }
              `);

        const pos = elm1.ref.getBoundingClientRect();

        expect(position.getInstance()).toStrictEqual({
          x: pos.left,
          y: pos.top,
        });

        expect(elm.id).toBe(elm1.id);
        expect(prevElm).toBeNull();
      });
    });
  });
});
