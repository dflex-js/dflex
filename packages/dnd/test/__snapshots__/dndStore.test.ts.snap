// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DnD Store Shallow test the store Element is initiated 1`] = `
Object {
  "id-1": CoreUtils {
    "animatedFrame": 1,
    "currentPosition": PointNum {
      "x": 450,
      "y": 114,
    },
    "depth": 0,
    "grid": PointNum {
      "x": 1,
      "y": 1,
    },
    "hasAttribute": Object {},
    "hasToTransform": false,
    "id": "id-1",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": true,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "offset": Object {
      "height": 50,
      "left": 450,
      "top": 114,
      "width": 170,
    },
    "order": Object {
      "parent": 0,
      "self": 0,
    },
    "ref": <div
      data-index="0"
    />,
    "translate": PointNum {
      "x": 0,
      "y": 0,
    },
  },
}
`;
